### DISCUSSION TITLE
Subclass of Button not working

### DISCUSSION THREAD

**Discussion by josephnarai** [AvaloniaUI/Avalonia#6704] [Q&A] [answered]
I have created a subclass of a Button, expecting to be able to simply use it in place of a normal Button, however it does not appear to be working.

Using the Avalonia VS code behind template

```
 public MainWindow()
        {
            InitializeComponent();
#if DEBUG
            this.AttachDevTools();
#endif
            StackPanel stackpanel = this.FindControl<StackPanel>("MyStackPanel");
            stackpanel.Children.Add(new Button()
            {
                Content = "Button 1"
            });
            stackpanel.Children.Add(new MyButton()
            {
                Content = "My Button 1"
            });
            stackpanel.Children.Add(new Button()
            {
                Content = "Button 2"
            });
            stackpanel.Children.Add(new MyButton()
            {
                Content = "My Button 2"
            });
        }
```

That is my MainWindow constructor.

And this

```
using Avalonia.Controls;
using System.Diagnostics;

namespace AvaloniaButtonTest
{
    internal class MyButton : Button
    {
        public MyButton()
        {
            Debug.WriteLine("My Button Constructor");
        }
    }
}
```

Is the MyButton class, inheriting from Button

I only see Button 1 and Button 2 on the stack panel, if I change MyButton to Button in the MainWindow constructor then I correctly see all 4 buttons...

I'm assuming I'm doing something silly, but can't work out why a simple subclass wouldn't work.

Please help!

Thanks in advance.

**Comment by MarkusKgit** [comment, marked as answer]
Avalonia somehow needs to know how to style your control. The easiest way to do that is to decorate your class with the `IStyleable` interface and implement the `StyleKey` property to use the default Button style:

```csharp
class MyButton : Button, IStyleable
{
    Type IStyleable.StyleKey => typeof(Button);
}
```
![Buttons](https://user-images.githubusercontent.com/13160892/136899402-26acbef2-1558-4513-835d-467844b9e653.png)

**Reply by josephnarai** [reply]
Awesome, thank you! - I thought it might need styling, but I could not find any reference as to how to style via C#. Is there any documentation on this kind of C# coding for Controls?

**Reply by MarkusKgit** [reply]
It is a bit hidden in a note [here ](https://docs.avaloniaui.net/docs/styling/selectors) (section OfType)

But you can also take inspiration from [here](https://github.com/AvaloniaUI/Avalonia/blob/df892e4a9ccd76ee54542d06fec4dd77772c2fdb/src/Avalonia.Diagnostics/Diagnostics/Controls/FilterTextBox.cs)

**Reply by josephnarai** [reply]
Yes, I found the first one, but there are no examples showing how you acutally use and set them!

For example, I'm going to want to change a few things in the default button style... I assume I need to make a new style in xaml based on the default button style and then somehow apply that new style to my button in C#... but I've not found any examples to show me how to achieve this.

Adding the IStyleable interface, didn't actually set a specific style to my new button, I'm aussuming it just tells avalonia that I am supposed to be styled like a button...

If you had any example code for what I'm trying to achieve, that would be greatly appreciated!

**Reply by MarkusKgit** [reply]
I think you need to make your own [Button style](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Button.xaml) in that case but I am not 100 % sure. But it says [here ](https://docs.avaloniaui.net/guides/developer-guides/comparison-of-avalonia-with-wpf-and-uwp#styling) that Avalonia doesn't have `Style.BasedOn` so that wouldn't work either.

**Reply by josephnarai** [reply]
Thanks, 

So I could make a completely new ButtonStyle from Scratch, that would be ok... I actually already did that in my previous (AdonisUI based) project.

How would I then assign that newly created style to the button from code behind? It's documented and simple in xaml, but I've not found a way to set it in C#.

**Reply by MarkusKgit** [reply]
You could move your own Button style in a separate axaml file (similar to the linked Button style). But instead of the Button selectors you would use your Button class. So `<Style Selector="Button">` would become `<Style Selector="local|MyButton">` with `local` being the namespace your Button resides in defined at the top (`xmlns:local="clr-namespace:AvaloniaButtonTest">`).
Then you can add that style to your global styles in App.axaml:
```
<Application.Styles>
    <FluentTheme Mode="Light"/>
    <StyleInclude Source="/MyButtonStyle.axaml"/>
</Application.Styles>
```

Again - maybe there is an easier way that I don't know of but thats how I would do it based on the documentation I could find.

**Reply by josephnarai** [reply]
Oh, I see, and that would then automatically apply the style to the button, if I set 

`Type IStyleable.StyleKey => typeof(MyButton);`

Thanks, I'll try that out and see how I go!

Much appreciated!

**Reply by MarkusKgit** [reply]
Actually you don't need the StyleKey in that case I believe

**Reply by josephnarai** [reply]
Oh, it will apply it because it knows the type! I see. I'll let you know how I go! 

Thanks!

**Reply by josephnarai** [reply]
Nearly there... 

I have created a new button style based on the Avalonia button

```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:local="clr-namespace:AvaloniaButtonTest">
  <Design.PreviewWith>
    <Border Padding="20">
      <StackPanel Spacing="20">
        <MyButton Content="Click Me!" />
        <MyButton Classes="accent" Content="Click Me!" />
      </StackPanel>
    </Border>
  </Design.PreviewWith>
  <Styles.Resources>
    <Thickness x:Key="ButtonPadding">8,5,8,6</Thickness>
  </Styles.Resources>
  <Style Selector="local|MyButton">
    <Setter Property="Background" Value="{DynamicResource ButtonBackground}" />
    <!--<Setter Property="BackgroundSizing" Value="OuterBorderEdge" />-->
    <Setter Property="Foreground" Value="{DynamicResource ButtonForeground}" />
    <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrush}" />
    <Setter Property="BorderThickness" Value="{DynamicResource ButtonBorderThemeThickness}" />
    <Setter Property="CornerRadius" Value="{DynamicResource ControlCornerRadius}" />
    <Setter Property="Padding" Value="{DynamicResource ButtonPadding}" />
    <Setter Property="HorizontalAlignment" Value="Left" />
    <Setter Property="VerticalAlignment" Value="Center" />
    <Setter Property="FontWeight" Value="Normal" />
    <Setter Property="FontSize" Value="{DynamicResource ControlContentThemeFontSize}" />
    <!--<Setter Property="UseSystemFocusVisuals" Value="{StaticResource UseSystemFocusVisuals}" />
    <Setter Property="FocusVisualMargin" Value="-3" />-->
    <Setter Property="Template">
      <ControlTemplate>
        <ContentPresenter x:Name="PART_ContentPresenter"
                          Background="{TemplateBinding Background}"
                          BorderBrush="{TemplateBinding BorderBrush}"
                          BorderThickness="{TemplateBinding BorderThickness}"
                          CornerRadius="{TemplateBinding CornerRadius}"
                          Content="{TemplateBinding Content}"
                          ContentTemplate="{TemplateBinding ContentTemplate}"
                          Padding="{TemplateBinding Padding}"
                          HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                          VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}" />
      </ControlTemplate>
    </Setter>
  </Style>

  <!-- PointerOverState -->
  <Style Selector="local|MyButton:pointerover /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="#FF00FF" />
    <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushPointerOver}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource ButtonForegroundPointerOver}" />
  </Style>

  <Style Selector="local|MyButton:pressed  /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource ButtonBackgroundPressed}" />
    <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushPressed}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource ButtonForegroundPressed}" />
  </Style>

  <Style Selector="local|MyButton:disabled /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource ButtonBackgroundDisabled}" />
    <Setter Property="BorderBrush" Value="{DynamicResource ButtonBorderBrushDisabled}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource ButtonForegroundDisabled}" />
  </Style>

  <Style Selector="local|MyButton.accent /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource AccentButtonBackground}" />
    <Setter Property="BorderBrush" Value="{DynamicResource AccentButtonBorderBrush}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource AccentButtonForeground}" />
  </Style>

  <Style Selector="local|MyButton.accent:pointerover /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource AccentButtonBackgroundPointerOver}" />
    <Setter Property="BorderBrush" Value="{DynamicResource AccentButtonBorderBrushPointerOver}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource AccentButtonForegroundPointerOver}" />
  </Style>

  <Style Selector="local|MyButton.accent:pressed  /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource AccentButtonBackgroundPressed}" />
    <Setter Property="BorderBrush" Value="{DynamicResource AccentButtonBorderBrushPressed}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource AccentButtonForegroundPressed}" />
  </Style>

  <Style Selector="local|MyButton, RepeatButton, ToggleButton">
    <Setter Property="RenderTransform" Value="none" />
    <Setter Property="Transitions">
      <Transitions>
        <TransformOperationsTransition Property="RenderTransform" Duration="0:0:.075" />
      </Transitions>
    </Setter>
  </Style>

  <Style Selector="local|MyButton:pressed, RepeatButton:pressed, ToggleButton:pressed">
    <Setter Property="RenderTransform" Value="scale(0.98)" />
  </Style>

  <Style Selector="local|MyButton.accent:disabled /template/ ContentPresenter#PART_ContentPresenter">
    <Setter Property="Background" Value="{DynamicResource AccentButtonBackgroundDisabled}" />
    <Setter Property="BorderBrush" Value="{DynamicResource AccentButtonBorderBrushDisabled}" />
    <Setter Property="TextBlock.Foreground" Value="{DynamicResource AccentButtonForegroundDisabled}" />
  </Style>
</Styles>
```

however, I'm getting the following build error:
`
Error	XAMLIL	Unable to resolve suitable regular or attached property CornerRadius on type AvaloniaButtonTest:AvaloniaButtonTest.MyButton Line 21, position 13.	AvaloniaButtonTest	C:\Dev\Zenso\AvaloniaButtonTest\AvaloniaButtonTest\MyButtonStyle.axaml	21	

`It is complaining that it doesn't know what a CornerRadius is... if I take out the corner radius, it works as expected and I can change styles elements!! 

Any clues as to why it can't resolve the CornerRadius property?

**Reply by josephnarai** [reply]
Just another note.

If I just import the default [Button style ](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Button.xaml) and add that to the project.

I still get the CornerRadius error, and can't build.

**Reply by MarkusKgit** [reply]
hmm it works for me ™ when I copy your style :-) I only had to change the `<Button ...>`s in the `Design.PreviewWith` section to `<local:MyButton ...>`s although that still compiled without the change.

**Reply by josephnarai** [reply]
That is very strange. I did fix the local:MyButton for the design preview and it works if I remove the CornerRadius.

I can't understand where this error is coming from either!

**Reply by josephnarai** [reply]
Ok. So it seems there was an Avalonia update!

Just updated via NuGet and it builds!

I only installed Avalonia two days ago! Very strange indeed, but at least it's working!

Thank you!
